crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -50,6 +50,10 @@ pub(super) struct SemanticIndexBuilder<'db> {
 
     /// Flow states at each `break` in the current loop.
     loop_break_states: Vec<FlowSnapshot>,
+    /// Are we in a state where a `break` statement is allowed?
+    /// This stack mirrors `scope_stack`.
+    inside_loop_stack: Vec<bool>,
```

Comment:
non-blocking comment: it might be slightly fewer allocations (and better reflect the invariant that the two stacks always have the same length) if we had something like this, rather than two parallel stacks:

```rs
enum LoopState {
    InLoop,
    NotInLoop,
}

pub(super) struct SemanticIndexBuilder<'db> {
    scope_stack: Vec<(FileScopeId, LoopState)>
}
```

But I'm also fine with what you have now!

Reply:
oops, I typed out this comment before I saw https://github.com/astral-sh/ruff/pull/14522#discussion_r1853798868 😆

Reply:
This is what I meant above by "I thought about amending the existing stack for scopes" :smile:. I'll check now what implications this has.

Reply:
It's better. Thanks for pushing me.